### PR TITLE
feat(android): expose installAPK(path) and ADB socket

### PR DESCRIPTION
**Pull request by mxschmitt**
https://github.com/microsoft/playwright/issues/1122

### CODE REVIEW

**Comment by pavelfeldman** [comment]
Few things:
- installApk, accept buffer and filename
- we don't read from filesystem on the server side, should be client
- you pass parameters unconditionally, but it is typical to pass different args while installing
- users might benefit from the actual open, so i wonder if it makes sense plumbing the connection object itself.

File: src/client/android.ts [hunk 1]
```diff
@@ -189,14 +193,31 @@ export class AndroidDevice extends ChannelOwner<channels.AndroidDeviceChannel, c
...
+    });
+  }
+
+  installApk(file: string | Buffer): Promise<void> {
+    return this._wrapApiCall('androidDevice.installApk', async () => {
+      const normalisedApk = await convertApkFile(file);
+      const adbSocket = await this.open(`shell:cmd package install -r -t -S ${normalisedApk.length}`);
+      debug('pw:android')('Writing driver bytes: ' + normalisedApk.length);
+      await adbSocket.write(normalisedApk);
+      const success = await new Promise(f => adbSocket.on('data', f));
```
Comment (pavelfeldman) [inline src/client/android.ts:214]: This response can be "Failed"

File: src/client/android.ts [hunk 2]
```diff
@@ -217,6 +238,30 @@ export class AndroidDevice extends ChannelOwner<channels.AndroidDeviceChannel, c
...
+    return (androidDevice as any)._object;
+  }
+
+  constructor(parent: ChannelOwner, type: string, guid: string, initializer: channels.AndroidAdbSocketInitializer) {
+    super(parent, type, guid, initializer);
+    this._channel.on('data', ({ data }) => this.emit('data', data)); // TODO(mxschmitt): use constant
+  }
+
+  async write(data: Buffer): Promise<void> {
+    return this._wrapApiCall('androidDevice.inputType', async () => {
```
Comment (pavelfeldman) [inline src/client/android.ts:253]: androidDevice.write

File: src/client/android.ts [hunk 3]
```diff
@@ -217,6 +238,30 @@ export class AndroidDevice extends ChannelOwner<channels.AndroidDeviceChannel, c
...
+      await this._channel.write({ data: data.toString('base64') });
+    });
+  }
+}
+
+async function convertApkFile(file: string | Buffer): Promise<Buffer> {
```
Comment (pavelfeldman) [inline src/client/android.ts:259]: nit: readApkFile

File: src/client/android.ts [hunk 4]
```diff
@@ -217,6 +238,30 @@ export class AndroidDevice extends ChannelOwner<channels.AndroidDeviceChannel, c
...
 
+export class AndroidAdbSocket extends ChannelOwner<channels.AndroidAdbSocketChannel, channels.AndroidAdbSocketInitializer> {
+
+  static from(androidDevice: channels.AndroidAdbSocketChannel): AndroidAdbSocket {
+    return (androidDevice as any)._object;
+  }
+
+  constructor(parent: ChannelOwner, type: string, guid: string, initializer: channels.AndroidAdbSocketInitializer) {
+    super(parent, type, guid, initializer);
+    this._channel.on('data', ({ data }) => this.emit('data', data)); // TODO(mxschmitt): use constant
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply src/client/android.ts:249]: You are emitting a string, not a buffer.

File: src/client/connection.ts [hunk 5]
```diff
@@ -154,6 +154,9 @@ export class Connection {
       case 'AndroidDevice':
         result = new AndroidDevice(parent, type, guid, initializer);
         break;
+      case 'AndroidAdbSocket':
```
Comment (pavelfeldman) [inline src/client/connection.ts:157]: That's not how sorting works :)

File: src/dispatchers/androidDispatcher.ts [hunk 6]
```diff
@@ -147,6 +156,15 @@ export class AndroidDeviceDispatcher extends Dispatcher<AndroidDevice, channels.
   }
 }
 
+export class AndroidAdbSocketDispatcher extends Dispatcher<AndroidAdbSocket, channels.AndroidAdbSocketInitializer> implements channels.AndroidAdbSocketChannel {
+  constructor(scope: DispatcherScope, socket: AndroidAdbSocket) {
+    super(scope, socket, 'AndroidAdbSocket', {}, true);
```
Comment (pavelfeldman) [inline src/dispatchers/androidDispatcher.ts:161]: I would think that you listen to the socket.

File: src/protocol/protocol.yml [hunk 7]
```diff
@@ -2083,6 +2083,19 @@ Android:
...
         timeout: number
 
+AndroidAdbSocket:
+  type: interface
+
+  commands:
+
+    write:
+      parameters:
+        data: string
```
Comment (pavelfeldman) [inline src/protocol/protocol.yml:2093]: data: binary

File: src/protocol/protocol.yml [hunk 8]
```diff
@@ -2083,6 +2083,19 @@ Android:
...
+
+  events:
+    data:
+      parameters:
+        data: string
```
Comment (pavelfeldman) [inline src/protocol/protocol.yml:2098]: data: binary

File: src/protocol/protocol.yml [hunk 9]
```diff
@@ -2275,12 +2288,22 @@ AndroidDevice:
...
+      returns:
+        socket: AndroidAdbSocket
+
     shell:
       parameters:
         command: string
       returns:
         result: string
 
+    installApk:
```
Comment (pavelfeldman) [inline src/protocol/protocol.yml:2303]: Looks like you don't need it.

File: src/server/android/android.ts [hunk 10]
```diff
@@ -297,6 +295,25 @@ export class AndroidDevice extends EventEmitter {
   }
 }
 
+export class AndroidAdbSocket extends EventEmitter {
```
Comment (pavelfeldman) [inline src/server/android/android.ts:298]: I don't think you need this class, just use backend socket.

File: src/server/android/android.ts [hunk 11]
```diff
@@ -338,7 +355,7 @@ Sec-WebSocket-Version: 13\r
     await new Promise(f => this._socket!.once('data', f));
 
     // Start sending web frame to receiver.
-    this._socket.on('data', data => this._receiver._write(data, 'binary', () => {}));
+    this._socket.on('data', data => this._receiver._write(data, 'binary', () => { }));
```
Comment (pavelfeldman) [inline src/server/android/android.ts:358]: Revert

File: src/server/android/android.ts (hunk 10 above)
Comment (mxschmitt) [reply src/server/android/android.ts]: Do you mean to move the logic into the BufferedSocketWrapper class since it implements the SocketBackend interface?

Comment (pavelfeldman) [reply src/server/android/android.ts]: What logic?

File: android-types-internal.d.ts [hunk 12]
```diff
@@ -46,6 +48,10 @@ export interface AndroidDevice<BrowserContextOptions, BrowserContext, Page> exte
   info(selector: AndroidSelector): Promise<AndroidElementInfo>;
 }
 
+export interface AndroidSocket extends EventEmitter {
+  write(data: Buffer): Promise<void>
```
Comment (pavelfeldman) [inline android-types-internal.d.ts:52]: declare on('data') here as well (see other examples)

File: src/dispatchers/androidDispatcher.ts [hunk 13]
```diff
@@ -147,6 +157,17 @@ export class AndroidDeviceDispatcher extends Dispatcher<AndroidDevice, channels.
...
+export class AndroidSocketDispatcher extends Dispatcher<SocketBackend, channels.AndroidSocketInitializer> implements channels.AndroidSocketChannel {
+  constructor(scope: DispatcherScope, socket: SocketBackend) {
+    super(scope, socket, 'AndroidSocket', {}, true);
+    socket.on(Events.AndroidSocket.Data, (data: string) => this._dispatchEvent('data', { data }));
+  }
+
+  async write(params: channels.AndroidSocketWriteParams, metadata?: channels.Metadata): Promise<void> {
+    await this._object.write(Buffer.from(params.data, 'base64'));
+  }
+}
```
Comment (pavelfeldman) [inline src/dispatchers/androidDispatcher.ts:169]: We are missing `close()`, every socket should have one.

File: src/server/android/backendAdb.ts [hunk 14]
```diff
@@ -111,7 +111,7 @@ class BufferedSocketWrapper extends EventEmitter implements SocketBackend {
     this._socket.on('data', data => {
       debug('pw:adb:data')(data.toString());
       if (this._isSocket) {
-        this.emit('data', data);
+        this.emit('data', data.toString('base64'));
```
Comment (pavelfeldman) [inline src/server/android/backendAdb.ts:114]: This doesn't sound right, my socket is a buffer socket.

File: src/dispatchers/androidDispatcher.ts [hunk 15]
```diff
@@ -147,6 +157,17 @@ export class AndroidDeviceDispatcher extends Dispatcher<AndroidDevice, channels.
   }
 }
 
+export class AndroidSocketDispatcher extends Dispatcher<SocketBackend, channels.AndroidSocketInitializer> implements channels.AndroidSocketChannel {
+  constructor(scope: DispatcherScope, socket: SocketBackend) {
+    super(scope, socket, 'AndroidSocket', {}, true);
+    socket.on(Events.AndroidSocket.Data, (data: string) => this._dispatchEvent('data', { data }));
```
Comment (pavelfeldman) [inline src/dispatchers/androidDispatcher.ts:163]: `{ data: data.toString('base64') }`

File: src/server/android/backendAdb.ts (hunk 14 above)
Comment (mxschmitt) [reply src/server/android/backendAdb.ts:114]: good catch, fixed.

File: src/dispatchers/androidDispatcher.ts (hunk 13 above)
Comment (mxschmitt) [reply src/dispatchers/androidDispatcher.ts:169]: good catch.

**Review by pavelfeldman — APPROVED** [review]

**Comment by pavelfeldman** [comment]
There are tiny nits that we can follow up on.

**Pull request merged**